ScientificCalculator/test/Spec.hs

```diff
@@ -0,0 +1,142 @@
+import qualified UnitTest as U
+
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import ParserExpr
+import qualified Eval
+import Test.Tasty
+import Test.Tasty.Hedgehog (testProperty)
+import ApplyFunctions
+
+
+
+genNumber :: Int -> Int -> Gen String
+genNumber a b = show <$> Gen.realFrac_ (Range.linearFrac (fromIntegral a) (fromIntegral b))
+
+genConstant :: Gen String
+genConstant = Gen.element ["pi", "e", "π"]
+
+
+nonRestrictedBinary = [ "+", "-", "*", "EE" ]
+
+restrictedBinary = ["/", "^", "Lny"]
+
+genOperator :: [String] -> Gen String
+genOperator = Gen.element
+
+nonRestrictedUnary = 
+  [ 
+    "sin", "cos", "tan", "exp", "sinh", "cosh", 
+    "tanh", "deg", "rad", "cbrt" , "x^2", "x^3"
+  ]
+
+restrictedUnary =  
+  [ 
+		"sqrt", "ln", "log10", "asin", "acos", 
+		"atan", "asinh", "acosh", "atanh",  "1/x"
+	]
+
+genUnaryFunction :: [String] -> Gen String
+genUnaryFunction = Gen.element 
+
+genSimpleExpr :: Int -> Int -> Gen String
+genSimpleExpr a b = Gen.choice [genNumber a b, genConstant]
+
+genAllExpr :: Int -> Gen String
+genAllExpr 0 = genSimpleExpr (-1000) 1000
+genAllExpr n = Gen.choice
+  [ 
+		genSimpleExpr (-1000) 1000
+  , do
+      func <- genUnaryFunction (nonRestrictedUnary ++ restrictedUnary)
+      expr <- genAllExpr (n - 1)
+      return $ func ++ "(" ++ expr ++ ")"
+  , do
+      expr1 <- genAllExpr (n `div` 2)
+      expr2 <- genAllExpr (n `div` 2)
+      op <- genOperator (nonRestrictedBinary ++ restrictedBinary)
+      return $ "(" ++ expr1 ++ " " ++ op ++ " " ++ expr2 ++ ")"
+  ]
```

Comment:
Good advice, but it was an implementation decision, so that I can check the parser and the evaluator at the same time.

Reply:
It's a poor implementation decision. Tests should strive to test one unit of functionality at a time. Your evaluator expects an expression, not its string representation, so generate an expression for it. Otherwise, you have two moving parts in the test, and when the test fails, you'll have to figure out which of them failed. To test your parser, you just show an expression and pass the string representation. If you really want to test your parser, showing should introduce some variations into how a string is formed.